### PR TITLE
Arc - improve producer wildcard type detection

### LINKED ISSUE #29836: ArC - improve the way a wildcard in a producer return/field type is detected
### Describe the bug

It seems that we don't search through the "nested" parameterized types.

Note that wildcards are not legal in a a producer return/field type. See also https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html#producer_field.

This is a follow-up of https://github.com/quarkusio/quarkus/pull/4756.

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

Replace the return type of `io.quarkus.arc.processor.TypesTest.Producer.produce()`; `List<? extends Number>` -> `List<Set<? extends Number>>`

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by manovotn**
Fixes #29836

### CODE REVIEW

**Comment by manovotn** [comment]
After discussion with @mkouba, we agreed that this scenario should instead of logging throw a `DefinitionException` based on the following [specification section](https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html#producer_method):

> If a producer method return type contains a wildcard type parameter or is an array type whose component type contains a wildcard type parameter, the container automatically detects the problem and treats it as a definition error.

(Same applies to producer fields OFC)

**Comment by manovotn** [comment]
@mkouba updated as per discussion.

Following exception is now thrown:

> javax.enterprise.inject.spi.DefinitionException: Producer method java.util.List<? extends java.lang.Number> produce() declared on class io.quarkus.arc.processor.TypesTest$Producer contains a parameterized type with a wildcard. This type is not a legal bean type according to CDI specification.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java [hunk 1]
```diff
@@ -445,19 +469,12 @@ static Set<Type> getTypeClosure(ClassInfo classInfo, AnnotationTarget producerFi
         } else {
             // Canonical ParameterizedType with unresolved type variables
             Type[] typeParams = new Type[typeParameters.size()];
-            boolean skipThisType = false;
             for (int i = 0; i < typeParameters.size(); i++) {
                 typeParams[i] = resolvedTypeParameters.get(typeParameters.get(i).identifier());
-                // this should only be the case for producers; wildcard is not a legal bean type
+                // for producers, wildcard is not a legal bean type and we want to skip that type
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java:474]: `and we want to skip that type` -> `and results in a definition error`?

File: independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/TypesTest.java [hunk 2]
```diff
@@ -67,15 +70,41 @@ public void testGetTypeClosure() throws IOException {
         String producersName = "produce";
         MethodInfo producerMethod = producerClass.method(producersName);
         // Object is the sole type
-        Set<Type> producerMethodTypes = Types.getProducerMethodTypeClosure(producerMethod,
-                dummyDeployment);
-        assertEquals(1, producerMethodTypes.size());
+        try {
+            Types.getProducerMethodTypeClosure(producerMethod, dummyDeployment);
+            fail("Should have thrown definition exception. Wildcard is not a legal bean type!");
+        } catch (DefinitionException e) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/TypesTest.java:76]: The `try` block is ok but `Assertions.assertThrows()` is sometimes more readable...

Comment (manovotn) [reply independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/TypesTest.java]: Fair enough, changed :)

**Review by mkouba — APPROVED** [review]

**Pull request merged**